mlflow/models/evaluation/default_evaluator.py

```diff
@@ -494,28 +537,57 @@ def _log_model_explainability(self):
             f: f2 for f, f2 in zip(self.feature_names, truncated_feature_names)
         }
 
-        sampled_X = shap.sample(self.X, sample_rows)
+        sampled_X = shap.sample(self.X, sample_rows, random_state=0)
+        if algorithm == "kernel":
+            background_X = shap.sample(self.X, sample_rows, random_state=3)
+        else:
+            background_X = None
 
         if isinstance(sampled_X, pd.DataFrame):
             # For some shap explainer, the plot will use the DataFrame column names instead of
             # using feature_names argument value. So rename the dataframe column names.
             sampled_X = sampled_X.rename(columns=truncated_feature_name_map, copy=False)
+            if background_X is not None:
+                background_X = background_X.rename(columns=truncated_feature_name_map, copy=False)
+        else:
+            # If sample_X is numpy array, convert to pandas dataframe.
+            sampled_X = pd.DataFrame(sampled_X, columns=truncated_feature_names)
+            if background_X is not None:
+                background_X = pd.DataFrame(background_X, columns=truncated_feature_names)
 
         if algorithm:
-            supported_algos = ["exact", "permutation", "partition"]
-            if algorithm not in supported_algos:
-                raise ValueError(
-                    f"Specified explainer algorithm {algorithm} is unsupported. Currently only "
-                    f"support {','.join(supported_algos)} algorithms."
+            if algorithm == "kernel":
+                kernel_link = self.evaluator_config.get("explainability_kernel_link", "identity")
+                if kernel_link not in ["identity", "logit"]:
+                    raise ValueError(
+                        "explainability_kernel_link config can only be set to 'identity' or "
+                        f"'logit', but got '{kernel_link}'."
+                    )
+                mode_or_mean_dict = _compute_df_mode_or_mean(self.X)
+                mode_or_mean_dict = {
+                    truncated_feature_name_map[k]: v for k, v in mode_or_mean_dict.items()
+                }
+                sampled_X = sampled_X.fillna(mode_or_mean_dict)
```

Comment:
What would happen if the imputation here is different from the imputation the model performs?

Reply:
For example, let's say we have a pipeline model that contains an imputer. What would happen if the imputer just fills NaN with `0`? Would this lead to incorrect feature importance?

Reply:
btw do we need imputation when the model contains an imputer?

Reply:
Yes, it might affect feature importance, But, I think this is a minor issue. Because the kernel explainer runs in approximate way, and we sample rows as the "background" data, the different sampling will also affect the result.

Reply:
@WeichenXu123 do we need to impute sampled_X when the model contains an imputer? SHAP doesn't work when sampled_X contains missing values?

Reply:
For KernelExplainer, shap can still work with NA values if we don't impute sampled_X and background_X.
Because KernelExplainer it won't transform the provided feature values before it feeds feature values into the model.
Rethink: do we still need imputation preprocessing for shap KernelExplainer ?

I link @BenWilson2 's old comment about this.
https://github.com/mlflow/mlflow/pull/5920#discussion_r879598221

Reply:
btw why did you think imputation is required?

Reply:
I think this brings up 3 separate discussion points...
1. Does the model handle missing data? 
    For models like XGBoost / LightGBM which are quite prevalent in usage for supervised learning tasks, an assumption by many is made that data is imputed. It is not. NaN / Null / NA values are all tagged as a 'special internal value' that can be split on (i.e. at a decision branch node, the logic may be 'col x is NA -> Left ; col x is not NA -> Right'. It doesn't fill in zeros or use mean values. 
    For linear models, imputation is absolutely needed and, if not explicitly handled by an imputer stage transformer (which is definitely a best practice but also one that is not universally performed by practitioners), will default to the modeling package implementation (consistency in behavior is not guaranteed).
2. SHAP handles missing values in a rather complex way. If the column under evaluation is evaluated as 'missing at random' (low counts of missing feature vector positions in a global sense), the local evaluation being performed will toggle that feature, resulting in a SHAP value of effectively 0 for that feature of that vector evaluation. Global missing throughout a vector position will essentially behave as a universal feature toggle and SHAP will report effectively 0 importance for the feature. Imputing sparse data will, naturally, change this behavior (but not by an appreciable amount). [I think it's low-risk on the accuracy side]
3. If the eventual goal is to support a broader application of efficient evaluation execution (supporting the full range of model-family-specific-implementations (i.e., TreeSHAP for tree-based models)) and to support arbitrary ML packages that may have dramatically different internal handling behavior for missing value elements, exposing a flexible config-driven logic for how to handle imputation and missing values to produce relevant feature importance results might be worthwhile.

Reply:
Updated, newest code changes:

1. Make imputation preprocessing for all kinds of Shap explainers, the imputation rule is: if we can confirm it is continuous values, use mean imputation, otherwise use mode imputation.

2. Add exception catching wrapper during running shap evaluation.